plutus-ir/src/Language/PlutusIR/Compiler/Datatype.hs

```diff
@@ -153,142 +219,137 @@ mkDatatypePatternFunctor d@(Datatype _ _ tvs _ _) = do
-
 -- Constructors
 
--- | Make the 'Def's for the constructors of a 'Datatype' with the given pattern functor.
-mkConstructorDefs :: forall m . MonadQuote m => Recursivity -> Type TyName () -> Datatype TyName Name () -> m [Def (VarDecl TyName Name ()) (PLC.Term TyName Name ())]
-mkConstructorDefs r pf (Datatype _ tn tvs _ constrs) =
-    forM constrs $ \c -> Def c <$> mkConstructor c
-        where
-            -- See note [Scott encoding of datatypes]
-            mkConstructor :: VarDecl TyName Name () -> m (PLC.Term TyName Name ())
-            mkConstructor constr = do
-                -- TODO: include type name
-                resultType <- liftQuote $ freshTyName () "r"
-
-                -- case arguments and their types
-                let caseTypes = fmap (constructorCaseType (PLC.TyVar () resultType)) constrs
-                caseArgNames <- forM [0..(length constrs -1)] (\i -> liftQuote $ freshName () $ strToBs $ "case" ++ show i)
-
-                -- TODO: this is inelegant, but it should never fail
-                let index = fromJust $ elemIndex constr constrs
-                let thisCase = caseArgNames !! index
-
-                -- constructor args and their types
-                let argTypes = constructorArgTypes constr
-                argNames <- forM [0..(length argTypes -1)] (\i -> liftQuote $ freshName () $ strToBs $ "arg" ++ show i)
-
-                let scottBody = mkScottConstructorBody r resultType (zip caseArgNames caseTypes) (zip argNames argTypes) thisCase (tyVarDeclName tn, pf)
-                pure $ mkIterTyAbs (fmap splitTyVarDecl tvs) scottBody
-
 -- See note [Scott encoding of datatypes]
--- | Make the body of a constructor.
-mkScottConstructorBody
-    :: Recursivity
-    -> TyName () -- ^ Name of the result type
-    -> [(Name (), PLC.Type TyName () )] -- ^ Names of the case arguments and their types
-    -> [(Name (), PLC.Type TyName () )] -- ^ Names of the constructor arguments and their types
-    -> Name () -- ^ Name of the current constructor
-    -> (TyName (), PLC.Type TyName ()) -- ^ A bound name and pattern functor
-    -> PLC.Term TyName Name ()
-mkScottConstructorBody r resultTypeName caseNamesAndTypes argNamesAndTypes thisConstructor (n, pf) =
-    let
-        -- data types are always in kind Type
-        resultKind = PLC.Type ()
-        -- See Note [Iterated abstraction and application]
-        -- c_i a_1 .. a_m
-        applied = foldl' (\acc a -> PLC.Apply () acc (PLC.Var () a)) (PLC.Var () thisConstructor) (fmap fst argNamesAndTypes)
-        -- \c_1 .. c_n . applied
-        cfuncs = mkIterLamAbs caseNamesAndTypes applied
+-- | Make a constructor of a 'Datatype' with the given pattern functor.
+-- @
+--     mkConstructor List (\(a :: *) -> forall (out_List :: *) . out_List -> (a -> List a -> out_List) -> out_List) Cons =
+--         /\(a :: *) . \(arg1 : a) (arg1 : List a) . /\(out_List :: *) . \(case_Nil : out_List) (case_Cons : a -> List a -> out_List) . case_Cons arg1 arg2
```

Comment:
Great, thank you! Let me suggest adding one another line:

```
mkConstructor List ListF Cons =
mkConstructor List (\(a :: *) -> forall (out_List :: *) . out_List -> (a -> List a -> out_List) -> out_List) Cons =
```

A typo: `arg1` is used twice: `\(arg1 : a) (arg1 : List a)`

You bind the `Cons` variable, but do not seem to use it? Does it represent any constructor of `List` or is just the `(:)` example?

Reply:
Right, I should expand `Cons` to the type we're given.

Reply:
Well, actually its main use here is to tell us we need `case_Cons` at the end.